File: JAVA/02_Recursion/Lecture14/targetSum.java
package Lecture14;

import java.util.*;

public class targetSum {

	public static void main(String[] args) {
		int[] arr = { 10, 20, 30, 40, 50, 60, 70 };
		ArrayList<ArrayList<Integer>> ans = tarSum(arr, 70, 0);

		System.out.println(ans);
	}

	// s p = starting point
	public static ArrayList<ArrayList<Integer>> tarSum(int[] arr, int tar, int sp) {

		if (sp == arr.length) {
			ArrayList<ArrayList<Integer>> bres = new ArrayList<ArrayList<Integer>>();
			if (tar == 0) {
				ArrayList<Integer> vk = new ArrayList<Integer>();
				bres.add(vk);
				return bres;
			}
			return bres;
		}

		ArrayList<ArrayList<Integer>> mres = new ArrayList<ArrayList<Integer>>();
		ArrayList<ArrayList<Integer>> rres1 = tarSum(arr, tar, sp + 1);
		ArrayList<ArrayList<Integer>> rres2 = tarSum(arr, tar - arr[sp], sp + 1);

		for (int j = 0; j < rres1.size(); j++) {
			mres.add(rres1.get(j));
		}
		for (int j = 0; j < rres2.size(); j++) {
			rres2.get(j).add(arr[sp]);
			mres.add(rres2.get(j));
		}

		return mres;

	}

}

File: C++/DS1/Lec13/targetSum.cpp
#include<iostream>
#include<vector>

using namespace std;

vector<vector<int>>* tarSum(int arr[], int tar, int sp, int n){

    if(sp == n){
        vector<vector<int>>* bres = new vector<vector<int>>();
        if(tar == 0){
            vector<int> vk;
            bres->push_back(vk);
        }
        return bres;
    }

    vector<vector<int>>* rres1 = tarSum(arr, tar - arr[sp], sp + 1, n);
    vector<vector<int>>* rres2 = tarSum(arr, tar, sp + 1, n);
    vector<vector<int>>* mres = new vector<vector<int>>();
    for(int i = 0; i < rres1 -> size() ; i++){
        (*rres1)[i].push_back(arr[sp]);
        (*mres).push_back((*rres1)[i]);
    }
    for(int i = 0; i < rres2 -> size() ; i++){
        (*mres).push_back((*rres2)[i]);
    }

    return mres;

}


int main(){
    int arr[] = {10, 20, 30, 40, 50, 60, 70};
    vector<vector<int>> * ans = tarSum(arr,70, 0, 7);
    //sp = starting point

    for(int i = 0; i < ans -> size(); i++){
        for(int j = 0; j < (*ans)[i].size(); j++){
            cout << (*ans)[i][j] << " " ;
         }
         cout << endl;
    }
    return 0;

}